### PR TITLE
remove unreachable error code `E0476`

### PR DESCRIPTION
**Pull request by Ezrashaw**
Remove `E0476` unreachable/undocumented/untested error code.

r? @GuillaumeGomez (I don't think you've noticed my other PRs for #61137 so I'll do more than just ping)

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me but since it's changing compiler's code, we'll need someone from their team.

r? @estebank

File: compiler/rustc_hir_analysis/src/coherence/builtin.rs [hunk 1]
```diff
@@ -386,8 +385,7 @@ pub fn coerce_unsized_info<'tcx>(tcx: TyCtxt<'tcx>, impl_did: DefId) -> CoerceUn
         (mt_a.ty, mt_b.ty, unsize_trait, None)
     };
     let (source, target, trait_def_id, kind) = match (source.kind(), target.kind()) {
-        (&ty::Ref(r_a, ty_a, mutbl_a), &ty::Ref(r_b, ty_b, mutbl_b)) => {
-            infcx.sub_regions(infer::RelateObjectBound(span), r_b, r_a);
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/coherence/builtin.rs:390]: Why is this line removed? This has side-effects.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/coherence/builtin.rs:390]: You could give it a different subregion origin, but this is required for soundness :sweat:

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/coherence/builtin.rs:390]: This code reproduces this error:

```rust
#![feature(coerce_unsized)]
#![feature(unsize)]

use core::marker::Unsize;
use core::ops::CoerceUnsized;

struct Wrapper<T>(T);

impl<'a, 'b, T, S> CoerceUnsized<&'a Wrapper<T>> for &'b Wrapper<S> where S: Unsize<T> {}
```

Comment (Ezrashaw) [reply compiler/rustc_hir_analysis/src/coherence/builtin.rs:390]: Ahh, my mistake, I assumed that this error code was no longer used. I'll close this PR and document/test it instead.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/coherence/builtin.rs:390]: It's alright -- in general, I would *never* remove logic just because you think it's dead code. Sometimes people just forget to add a UI test for an error code.

Replace it with a `delay_span_bug` or `bug!()` if necessary, but unless you can statically prove that something is unreachable, you're very likely to introduce unsoundness or other subtle bugs into the compiler.

Comment (Ezrashaw) [reply compiler/rustc_hir_analysis/src/coherence/builtin.rs:390]: Hmm, I was kind of naive. I assumed that untested error codes (basically all of them are like this) were from old Rust versions and had never been formally removed.
